Fixes #135: Handle application configuration using `QSettings`

mauricio1802 opened the pull request:
Fixes #135 
Fixes #284 

**Changes:**
- Change the approach used for the app settings
- Create a Config system in the backend allowing to plugins to define their own settings in a plugable way
- QSettings is used for accomplish the persisten over multiple platform
- Add `make test-docker` target to run commands inside `thereciepe/qt:linux` container
- Run `make test-docker` on Travis CI since `QSettings` tests rely on `therecipe/qt`

**Does this change need to mentioned in CHANGELOG.md?**
 no

**Requires testing**
yes

mauricio1802 committed:
[config] Create ConfigSection and Option types

mauricio1802 committed:
[config] Implement RegisterSection method for ConfigManager type

mauricio1802 committed:
[config] Remove ConfigSection type and Add config.go

mauricio1802 committed:
[config] Add GetValues method to SectionManager

mauricio1802 committed:
[config] Change init and LoadWalletEnvs functions to fit with new config

mauricio1802 committed:
[config] Change Option type to allow multiple levels of groups

mauricio1802 committed:
[config] Change SectionManager methods to allow multi level sections

mauricio1802 committed:
[config] Fix methods in skycoin plugin to work with config

mauricio1802 committed:
[config] Create multi platform path to use it as default in wallet source

mauricio1802 committed:
Merge branch 'fiberDevelop' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

mauricio1802 committed:
[config] Remove debug lines

mauricio1802 committed:
[config] Update init function of skycoin package to use new Config approach

mauricio1802 committed:
Merge branch 'fiberDevelop' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

mauricio1802 committed:
[plugin] Move creation and register of skycoin plugin to skycoin/main.go

mauricio1802 committed:
[config] Move Register of skycoin plugin to src/coin/skycoin/models/main.go

mauricio1802 committed:
Merge commit '608c4de' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

mauricio1802 committed:
[plugin] Rewrite LoadWalletEnv function

stdevCrow committed:
Rename the function `edit` to `setValue` and add a function `getValue` to complement it

stdevCrow committed:
Read the default settings with the new functions implemented in the `ConfigManager`

mauricio1802 committed:
[config] Add setValue and getValue methods to ConfigManager

mauricio1802 committed:
[test] Move model test to src/coin/skycoin/models/main_test.go

mauricio1802 committed:
[config] Fix error spliting path

mauricio1802 committed:
Merge branch 'fiberDevelop' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

olemis commented:
@stdevMauricio1802 is this ready to be merged ?

olemis reviewed (changes requested):
Needs test cases

hsequeda committed:
[ui] refs #135 Reimplemented Setting page to use the new model.

stdevCrow committed:
Merge branch 'stdevMauricio1802_t135_Handle_application_configuration_using_QSettings' of https://github.com/simelo/FiberCryptoWallet into base/stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

stdevCrow commented on src/models/config.go:
```diff
@@ -1,85 +1,179 @@
+	_             func(string) string         `slot:"getValue"`
+	_             func() []string             `slot:"getSections"`
+	_             func(string) *ConfigSection `slot:"getSection"`
 }
```
We need a way to get the default values. Something like a `getDefault("skycoin/walletSource/1/Source")` should return `"/home/username/.skycoin/wallets"`.

olemis replied:
@stdevMauricio1802 already added that feature . can this conversation be marked as resolved ?

stdevCrow commented:
@stdevMauricio1802 `Apply` and `Discard` settings work as expected, but `Restore defaults` don't. Please check the requested changes. Once done we can merge and close this PR.

stdevCrow committed:
Fix the organization domain: `simelo.tech.org => simelo.tech`

stdevCrow committed:
Properly bind settings with the backend
Implement functions to **save** and **get saved** settings
Minor grammar and code layout improvements

stdevCrow committed:
Clarify that the wallet path configuration is for local environment only

stdevCrow committed:
Remove redundant label `"Local wallet path"`

stdevCrow committed:
Remove redundant label `Node Url`

stdevCrow committed:
Fix weird `1` pixel offset of the `MenuBar` respect to the top of the application

stdevCrow committed:
Replace the `SwitchDelegate` for a simple `Switch` with two `Label`, one in each side indicating what means the switch in the respective position

stdevCrow committed:
Refactor some blocks of code into an independent function.
Add an `updateFooterButtonsStatus` function to update the footer buttons status ;)
Rename some functions and properties.
Simplify code layout.

stdevCrow committed:
Merge branch 'develop' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

mauricio1802 committed:
[models] Add getDefaultValue method to ConfigManager

olemis committed:
Merge remote-tracking branch 'simelo_alt/stdevMauricio1802_t135_Handle_application_configuration_using_QSettings' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

olemis committed:
[ci] refs #135 - Fix indentation error in Travis script

olemis committed:
[ci] refs #135 - Clean source before running tests on Travis CI

olemis committed:
[util] refs #135 - Replace assert => require in util package cipher tests

olemis committed:
[ci] refs #135 - Run tests inside Docker container during Travis builds

olemis committed:
[ci] refs #135 - Add (missing) therecipe/qt Docker image name in make test-docker

olemis committed:
[ci] refs #135 - Append default arch name to Docker image name in make test-docker

olemis committed:
[ci] refs #135 - Add src step in Makefile GOPATH_SRC var

stdevMac committed:
Merge remote-tracking branch 'origin/develop' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

olemis committed:
[ci] refs #135 - Run commands inside Docker container with make run-docker

Use it in Travis script

olemis committed:
[ci] refs #135 - Determine therecipe/qt GOPATH using docker inspect

olemis committed:
[ci] refs #135 - Implement DOCKER_GOPATH_SRC as run-docker target specific variable

olemis committed:
[ci] refs #135 - Append project subpath to Docker GOPATH to mount volume in container created for run-docker

olemis committed:
[ci] refs #135 - DOCKER_GOPATH_SRC => DOCKER_GOPATH in make docker-run

olemis committed:
[ci] refs #135 - Switch to project source folder inside Docker container prior to running make run-docker command

Project source folder is the one mounted internally under container GOPATH

olemis committed:
[ci] refs #135 - For Linux builds run make install-coveralls inside Docker container

olemis committed:
[ci] refs #135 - Use GOPATH (rather than HOME) to run goveralls in make test-cover-travis

stdevCrow committed:
Merge branch 'develop' of https://github.com/fibercrypto/FiberCryptoWallet into base/stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

stdevCrow committed:
Update the `default[...]` properties to get the real default settings from the backend
Add a `restoreDefaultSettings` function and add its call appropriatelly
Not need to save the current settings when accepting the `dialogConfirmation` dialog
Make the `dialogConfirmation` get focus only when visible

stdevCrow committed:
Get the organization and application's name from the official values instead of hard-cored ones

olemis committed:
[ci] refs #135 - Disable test coverage on Travis CI Linux builds

olemis committed:
Merge remote-tracking branch 'simelo_alt/stdevMauricio1802_t135_Handle_application_configuration_using_QSettings' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

olemis committed:
[ci] refs #284 - Use simelotech/qt-test for make run-docker

... since it includes a newer git version

olemis committed:
[ci] refs #135 - Install unzip on Travis Windows builds

stdevMac committed:
[lint] refs #135 Resolve lint issues

stdevMac committed:
Merge remote-tracking branch 'simelo/stdevMauricio1802_t135_Handle_application_configuration_using_QSettings' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

stdevMac committed:
Run qtmoc desktop in windows build before tests

stdevMac committed:
[qt] refs #135 Install qt in windows

stdevMac committed:
[ci] refs #135 Add travis_wait to qt download

stdevMac committed:
Install qt in windows using config script

stdevMac committed:
[ci][win] refs #135 add travis_wait to qt installation

stdevMac committed:
[qt] refs #135 Run qtsetup propertly

stdevMac committed:
[ci] refs #135 Set travis_wait time to 50 min

stdevMac committed:
[ci] refs #135 Test uname_s in travis windows

stdevMac committed:
[ci] refs #135 just run uname_s in travis windows build

stdevMac committed:
[ci] refs #135 remove # in choco install make in travis

stdevMac committed:
[make][ci] refs #135 Add rule for travis windows installation

stdevMac committed:
[ci] refs #135 Restore travis and add build on windows directly

stdevMac committed:
[ci] refs #135 Add cache to travis

stdevMac committed:
[make][ci] refs #135 set GO111MODULE=off in windows build

stdevMac committed:
[ci] remove gzip installation in travis

stdevMac committed:
[make] refs #135 remove | true in qtsetup

stdevMac committed:
[ci] refs #135 Add Appveyor to project

stdevMac committed:
[ci] refs #135 mv appveyor.yml .appveyor.yml

stdevMac committed:
Run using bash -c

stdevMac committed:
[ci] refs #135 Add enviroments var to appveyor

stdevMac committed:
[ci] Remove qt installation in appveyor

stdevMac committed:
[ci] run ls c:/qt

stdevMac committed:
[ci] refs #135 Use mingw as default shell in appveyor

stdevMac committed:
Set qt_version to 5.13

stdevMac committed:
[make] refs #135 set | true in qtsetup

stdevMac committed:
[ci] refs Install other qt version

stdevMac committed:
[ci] refs#135 Using mingw shell from qt installation

stdevMac committed:
[ci] Install therecipe/qt using -tags=no_envs

stdevMac committed:
Update vendor

stdevMac committed:
[changelog] Add address book

stdevMac committed:
[ci] refs #135 Remove appveyor scripts

stdevMac committed:
Merge remote-tracking branch 'simelo/stdevMac_update_vendor_and_changelog' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings

stdevMac committed:
Revert "Merge remote-tracking branch 'simelo/stdevMac_update_vendor_and_changelog' into stdevMauricio1802_t135_Handle_application_configuration_using_QSettings"

This reverts commit 2117f31ae5124965ad1e3ddd8ab8414f9c4e9e38, reversing
changes made to 761fcc9fc845889fb163da455a4ef426af6c451c.
